src/Microsoft.ML.Data/Data/SlotCursor.cs

```diff
@@ -0,0 +1,140 @@
+        /// </summary>
+        [BestFriend]
+        private protected bool IsGood => _state == CursorState.Good;
+
+        [BestFriend]
+        private protected SlotCursor(IChannelProvider provider)
+        {
+            Contracts.AssertValue(provider);
+            Ch = provider.Start("Slot Cursor");
+            _state = CursorState.NotStarted;
+        }
+
+        /// <summary>
+        /// The slot index. Incremented by one when <see cref="MoveNext"/> is called and returns <see langword="true"/>.
+        /// When initially created, or after <see cref="MoveNext"/> returns <see langword="false"/>, this will be <c>-1</c>.
+        /// </summary>
+        public abstract int SlotIndex { get; }
+
+        /// <summary>
+        /// Advance to the next row. When the cursor is first created, this method should be called to
+        /// move to the first slot. Returns <see langword="false"/> if there are no more slots.
+        /// </summary>
+        public abstract bool MoveNext();
+
+        /// <summary>
+        /// The slot type for this cursor. Note that this should equal the
+        /// <see cref="ITransposeSchema.GetSlotType"/> for the column from which this slot cursor
+        /// was created.
+        /// </summary>
+        public abstract VectorType GetSlotType();
+
+        /// <summary>
+        /// A getter delegate for the slot values. The type <typeparamref name="TValue"/> must correspond
+        /// to the item type from <see cref="ITransposeSchema.GetSlotType"/>.
+        /// </summary>
+        public abstract ValueGetter<VBuffer<TValue>> GetGetter<TValue>();
+
+        public virtual void Dispose()
+        {
+            if (_state != CursorState.Done)
+            {
+                Ch.Dispose();
+                _state = CursorState.Done;
+            }
+        }
+
+        /// <summary>
+        /// For wrapping another slot cursor from which we get <see cref="SlotIndex"/> and <see cref="MoveNext"/>,
+        /// but not the data or type accesors.
+        /// </summary>
+        [BestFriend]
+        internal abstract class SyncBase : SlotCursor
+        {
+            private readonly SlotCursor _root;
+
+            public SyncBase(IChannelProvider provider, SlotCursor cursor)
+                : base(provider)
+            {
+                Contracts.AssertValue(cursor);
+                // If the input is itself a sync-base, we can walk up the chain to get its root,
+                // thereby making things more efficient.
+                _root = cursor is SyncBase sync ? sync._root : cursor;
+            }
+
+            public override bool MoveNext()
+                => _root.MoveNext();
+
+            public override int SlotIndex => _root.SlotIndex;
+        }
+
+        /// <summary>
+        /// A useful base class for common <see cref="SlotCursor"/> implementations.
+        /// </summary>
+        [BestFriend]
+        internal abstract class RootBase : SlotCursor
```

Comment:
>RootBase [](start = 32, length = 8)

I might have called the other SlotCursorBase, and this one SlotCursor. Not sure of the relationship between Root and Slot.

Reply:
Just as the abstract class for `RowCursor` is named just `RowCursor` and not `RowCursorBase`, I would prefer hat the abstract class for a slot cursor just be `SlotCursor` rather than `SlotCursorBase`. But I can rename them and add comments to clarify.